### PR TITLE
fix: Adapt to PR coq-community/docker-coq#47

### PR DESCRIPTION
**Pull request by erikmd**
~~Can be merged and released after https://github.com/coq-community/docker-coq/pull/7 is deployed~~

~~This is a backward-compatible migration, so if there were no new release of the action, `coq-community/docker-coq-action@v1.2.2` will continue to work as is with existing configurations until December 14, since Docker-Coq's "default tags" (which don't specify an ocaml version) will be retained as dual-switch up to December 14.~~

**Edit:** Adapt to https://github.com/coq-community/docker-coq/pull/47

### CODE REVIEW

**Comment by Zimmi48** [comment]
Oh, minimal is deprecated? This is a surprise to me.

**Comment by Zimmi48** [comment]
BTW, in the end there was no announcement yet for the change in Docker-Coq single-switch vs double-switch, right? Should the end of the transition period still be on Dec 14th then?

**Comment by erikmd** [comment]
@Zimmi48
> BTW, in the end there was no announcement yet for the change in Docker-Coq single-switch vs double-switch, right? Should the end of the transition period still be on Dec 14th then?

Hi, I wanted to announce it yesterday, but I've been busy. Soon anyway.

And which migration period do you think it is reasonable? 7 days maybe?

Finally, if you agree with the "backward-compatible" solution we discussed just now, here is the plan:

* Continue to follow semantic versioning & Don't bump the docker-coq-action to v2
* Keep `ocaml_version: "minimal"` working, only deprecate it so that we can remove it when the v2 will land
* Add `ocaml_version: "default"` that does not hardcode an ocaml version but rely on the Docker-Coq default;
    that image will change at the end of the migration (e.g. coqorg/coq:8.12 is currently two-switch, but we will have coqorg/coq:8.12 = coqorg/coq:8.12-ocaml-4.07-flambda afterwards); from the point of view of docker-coq-action users, this will only change the "default ocaml version" for users that did not specify a given version (configurations with `ocaml_version: "minimal"` won't be impacted)
* Update the documentation accordingly.

Let me know if you are OK with this plan, and I'll implement it very soon in this PR :)

Finally, a suggestion regarding the deprecation of `"minimal"`: we could also print a warning in the log to document this further.

**Comment by Zimmi48** [comment]
I agree with this plan. 7 days sound right.

> Finally, a suggestion regarding the deprecation of `"minimal"`: we could also print a warning in the log to document this further.

Yes, it makes sense for a deprecated field.

The question that remains to be decided if whether to deprecate but preserve the semantics of `"minimal"` or to deprecate and change its semantics at the same time (to be synonymous of `"default"`).

The latter would not be strictly compatibility-preserving, but we can decide depending on the usages that were made of it.

I just reviewed the 8 usages found with: https://github.com/search?q=docker-coq-action+minimal+language%3AYAML&type=code

Most come directly from a generation from a previous version of the coq-community templates. Note that the coq-community template used an explicit `"minimal"` from May 1st, when it was introduced (coq-community/templates#26), to June 26th (coq-community/templates#59).

The only exceptions are:
- The configuration of dblib, which tests as many OCaml versions as possible, but since this is a pure Coq library, there is a PR open proposing to switch to standard template-based, single-OCaml-version testing (coq-community/dblib#11).
- The configurations of Proof General, Hierarchy Builder and Coq-Elpi, which all seem to be inspired by the templates.

Out of these, only Coq-Elpi is a plugin. It is currently tested with Coq 8.12 and OCaml `"minimal"`. If the semantics of `"minimal"` change to mean `"default"` and thus the plugin is not tested against OCaml 4.05 anymore (but 4.07 instead), this could be seen as a problem by its author (@gares), though it is easy to replace this `"minimal"` by an explicit `"4.05"` in this particular case.

WDYT?

**Comment by erikmd** [comment]
Hi @Zimmi48, thanks for your comments.
Regarding the semantics of `"minimal"` (as I also proposed to introduce `"default"` and you agreed with this new value), I actually see two sensible choices:

* add the "minimal business logic" to keep supporting "the minimal supported version in Docker-Coq"
* or remove that value "minimal" (but I guess we both think it'd be better not to do it for v1)

Indeed, having a `"minimal"` option that does not match its intended semantics would look a bit confusing to me, and having two values `minimal = default` with the same semantics (even if one is deprecated) seems a bit redundant.

Anyway, I agree with your concern that there should be the least possible business logic / ad-hoc code in docker-coq-action to "compute" the desired version! but I guess this can be done in a somewhat "future-proof" way (without needing to update the action as soon as a Coq is released that has the same minimal supported ocaml version as that of the previous release), as long as we we don't bump to docker-coq-action v2...

**Comment by Zimmi48** [comment]
If you are not comfortable with `minimal = default`, I'm fine with keeping `minimal` with some specific business-logic for it.

**Comment by erikmd** [comment]
Hi @Zimmi48, just FYI the "minimal" deprecation warning is implemented (and I guess it's in line which part of this Thursday's discussion on promoting a better version than 4.05.0!)
to have a look on the final rendering, see the [Annotations summary](https://github.com/erikmd/docker-coq-github-action-demo/actions/runs/414274179) (the 2nd annotation) − or the [GHA log itself](https://github.com/erikmd/docker-coq-github-action-demo/runs/1534261599?check_suite_focus=true#step:4:174)
(as for the rest of the PR implem: soonish)

**Comment by Zimmi48** [comment]
I'm surprised it doesn't also appear in the files changes tab as the Coq annotations themselves: https://github.com/erikmd/docker-coq-github-action-demo/pull/8/files

Apart from this, LGTM.

File: README.md
Comment (Zimmi48) [inline README.md]: Doc will need an update.

**Comment by LasseBlaauwbroek** [comment]
I suspect that this PR may have been forgotten to be merged together with the single-opam-switch change. Since then, `ocaml-version` no longer has an effect. Is there a reason this is delayed? Thanks.

**Comment by LasseBlaauwbroek** [comment]
I now see that the single-switch change may have been expected to be backward compatible. However, I need to move from 4.09 to 4.10 due to this change, and the CI is not properly selecting the 4.10 image. See here for example: https://github.com/coq-tactician/coq-tactician-stdlib/runs/1554834940

**Comment by erikmd** [comment]
@LasseBlaauwbroek my bad… the fact that this PR has been delayed induced that ocaml_version was ignored for the newly avaialble swiches!
We cannot merge #39 just now because it leads to other changes (the deprecation of "minimal", the addition of "default", doc to be finalized) but I'll push an independent fix right away…

**Comment by erikmd** [comment]
Hi @LasseBlaauwbroek, the issue is fixed now in [v1.2.3](https://github.com/coq-community/docker-coq-action/releases/tag/v1.2.3), thanks again for your report!

**Comment by LasseBlaauwbroek** [comment]
Thanks!

**Comment by erikmd** [comment]
@Zimmi48 could you proofread the docs part of this PR? (README.md and error messages only)

Here is the [link to the rendered `README.md`](https://github.com/coq-community/docker-coq-action/blob/single-switch/README.md) (only the `coq_version` and `ocaml_version` sections changed).

Regarding the error messages [1](https://github.com/coq-community/docker-coq-action/pull/39/files#diff-6f9d41d046756f0ddc2fcee0626bdb50100d12b88f293734eff742818e03efa2R123-R126) and [2](https://github.com/coq-community/docker-coq-action/pull/39/files#diff-6f9d41d046756f0ddc2fcee0626bdb50100d12b88f293734eff742818e03efa2R148-R151), they can be seen, rendered, as [GHA annotations](https://github.com/coq-community/docker-coq-action/actions/runs/2524288054):

2022-06-20_00-43-08_Screenshot_dev_minimal

2022-06-20_00-43-28_Screenshot_coq-native_ocaml-not-default

I plan to **merge this PR on next Monday 2022-06-20 and do a minor release**.
Next, remove the deprecated value of `"minimal"` the week after, on 2022-06-27 AOE.

**Comment by erikmd** [comment]
FTR, it seems that all Docker-Coq images experience a coq reinstall at runtime… probably because of [this change](https://github.com/ocaml/opam-repository/pull/21564), so I'm going to rebuild all Docker-Coq images… Done :heavy_check_mark:

**Comment by erikmd** [comment]
FWIW, to date there's only two YAML configs that use `"minimal"`:
https://github.com/search?q=docker-coq-action+minimal+language%3AYAML&type=code
so I'll open a PR to suggest replacing `"minimal"` with `"4.05"` *after* this PR is merged and released.

File: README.md
Comment (Zimmi48) [inline README.md]: What is the reason for providing `4.07` and `4.09` without the flambda option? Are they compatibility aliases? If that's the case, I get it for `4.07` but not for `4.09` given that it wasn't in the list before.

**Review by Zimmi48 — APPROVED** [review]

File: README.md
Comment (erikmd) [reply README.md]: Hi @Zimmi48, good catch:
On the one hand, these values are here just because at the time the contents of the `coq-native` Docker images were discussed, it was decided to only provide ocaml switches *without* flambda when `-native-compiler yes` is enabled.

But on the other hand, I agree that this may sound confusing, and moreover, these values of **don't make sense for docker-coq-action's `ocaml_version`** because it is precisely documented elsewhere that when we use a `coq-native` image, `ocaml_version` should be **`"default"`** or the empty string.
So, I'll just remove this noise from this very line of the README.md.
Thanks again Théo ! :+1:

**Comment by erikmd** [comment]
Let's wait for a newest CI run, then I'll merge and release this important patch tonight.

**Pull request merged**
